### PR TITLE
Fix attribute completion for expressions with comparison operators

### LINKED ISSUE #14897: Tab completion fails with comparison operators
### Description

When using tab completion in IPython, it fails to correctly identify the object type when the cursor is after a comparison operator.

For example:

`x = y = 'a'` works correctly: `x.<tab>` shows string methods
But `x == y.<tab>` incorrectly shows boolean methods instead of string methods

### Expected behavior

`x == y.<tab>` should complete with string methods for y

### System Information

IPython version: 9.3.0
Operating system: Ubuntu

### PR DESCRIPTION
**Pull request by Darshan808**
### Description

Modified `_attr_matches()` to check if an expression contains comparison operators `(=, ==, !=, etc.)` outside of parentheses. If so, it extracts the right-hand side of the operator for evaluation, providing appropriate completions.
For parenthesized expressions like `(x==y).<tab>`, the code preserves the entire expression to ensure it's evaluated as a bool.

Fixing completions fully is quite tricky. The current implementation still has flaws, especially in cases where comparison operators are involved. For example:
```python
{"==", "abc"}.<tab>
```
In this case, dictionary methods are not shown because the logic incorrectly separates the expression at the `"=="` operator.

While the current solution resolves some major issues, it also introduces new, albeit minor, ones.

Feedback on better approaches or alternative strategies is highly appreciated.

### CODE REVIEW

File: IPython/core/completer.py [hunk 1]
```diff
@@ -1142,6 +1142,13 @@ def _attr_matches(
             return [], ""
         expr, attr = m2.group(1, 2)
 
+        operators = ["=", "==", "!=", ">=", "<=", ">", "<"]
+        # Split by the operator and take the right side
+        if not (expr.startswith("(") and expr.endswith(")")):
+            for op in operators:
+                if op in expr:
+                    expr = expr.split(op)[-1]
```
[the remark this answers is not in the document]
Comment (krassowski) [reply IPython/core/completer.py:1150]: What about something like:

```python
def _strip_code_before_operator(code: str) -> str:
    operators = ["=", "==", "!=", ">=", "<=", ">", "<"]
    parens = {'(', '[', '{'}
    tokens = _parse_tokens(code)
    encountered_operator = False
    after_operator = None
    nesting_level = 0
    for t in tokens:
        if t.string in parens:
                nesting_level += 1
        elif t.string in parens:
                nesting_level -= 1
        elif t.type == tokenize.OP and t.string in operators and nesting_level == 0:
            encountered_operator = True
            after_operator = []
        elif encountered_operator:
            after_operator.append(t.string)
    if encountered_operator:
        return ''.join(after_operator)
    return code
```

```suggestion
        # Split by the operator and take the right side
        expr = _strip_code_before_operator(op)
```

File: IPython/core/completer.py [hunk 2]
```diff
@@ -1142,6 +1142,13 @@ def _attr_matches(
             return [], ""
         expr, attr = m2.group(1, 2)
 
+        operators = ["=", "==", "!=", ">=", "<=", ">", "<"]
+        # Split by the operator and take the right side
+        if not (expr.startswith("(") and expr.endswith(")")):
```
Comment (krassowski) [inline IPython/core/completer.py:1147]: How would this behave for `[a == b].`?

File: IPython/core/completer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (krassowski) [reply IPython/core/completer.py:1150]: Actually, I maybe we should not use `_parse_tokens` which ignores `tokenize.TokenError` and instead handle that error here directly so that `expr` is not changed if the code could not be tokenized, so that `_trim_expr` can handle it.

**Comment by Darshan808** [comment]
@krassowski 
Thanks for the feedback. It's working well for the cases I’ve tried so far. Do you happen to have any edge cases in mind that I should test against?

**Comment by krassowski** [comment]
> Do you happen to have any edge cases in mind that I should test against?

I thought about checking against maths operators (`+  -  *  /  //  %  ** @`), bitwise operators (`&  |  ^  <<  >>`) and assignment operators (`+=  -=  *=  /=  //=  %=  **= @= &=  |=  ^=  >>=  <<=`). 

Given `a, b = 1, 2.2` It looks like `a + b.<tab>` does not work currently, even though it used to work in older IPython versions; similar with other maths operators. Bitwise operators seem to work fine, but I only tested integers.

File: tests/test_completer.py [hunk 3]
```diff
@@ -654,7 +655,34 @@ def _(line, cursor_pos, expect, message, completion):
...
+                15,
+                ".upper",
+                "Should have completed on `x.upper() == y.`: %s",
+                Completion(15, 15, "upper"),
+            )
+            _(
+                "(x.upper() == y.",
+                16,
+                ".upper",
+                "Should have completed on `x.upper() == y.`: %s",
```
Comment (krassowski) [inline tests/test_completer.py:669]:
```suggestion
                "Should have completed on `(x.upper() == y.`: %s",
```

**Review by krassowski — COMMENTED** [review]
> Do you think it might be a good idea to strip all operators except the dot operator? It seems to pass all the cases I can think of.

Makes sense. Can you add the `a + b` to test cases, and then run darker to make the CI pass?

**Review by krassowski — APPROVED** [review]
Thank you @Darshan808!
